### PR TITLE
Register parameter execution modes.

### PR DESCRIPTION
**Pull request by ghislainfourny**

### CODE REVIEW

File: src/main/java/org/rumbledb/compiler/VisitorHelpers.java [hunk 1]
```diff
@@ -53,9 +81,13 @@ public static DynamicContext createDynamicContext(Node node, RumbleRuntimeConfig
 
     private static void setLocalExecutionForUnsetUserDefinedFunctions() {
         try {
-            for (
-                FunctionIdentifier functionIdentifier : Functions
+            List<FunctionIdentifier> unsetFunctionIdentifiers = new ArrayList<>();
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/compiler/VisitorHelpers.java:84]: Avoid a concurrent modification exception.

File: src/main/java/org/rumbledb/compiler/VisitorHelpers.java [hunk 2]
```diff
@@ -20,16 +23,33 @@ public static void resolveDependencies(Node node, RumbleRuntimeConfiguration con
...
+
+    public static void populateStaticContext(Node node, RumbleRuntimeConfiguration conf) {
+        printTree(node, conf);
         StaticContextVisitor visitor = new StaticContextVisitor();
         visitor.visit(node, null);
 
+
         visitor.setVisitorConfig(VisitorConfig.staticContextVisitorIntermediatePassConfig);
         int prevUnsetCount = Functions.getUserDefinedFunctionIdentifiersWithUnsetExecutionModes().size();
+        printTree(node, conf);
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/VisitorHelpers.java:45]: I think the printTree() function should do one thing that is to actually print the tree. I believe moving `if (conf.isPrintIteratorTree()) {` check into that function does more harm than good in terms of readability. As such, the code has the illusion that it always prints the iterator tree and only when you check the printTree() function you get the idea that it is in fact not printed thanks to the configuration.

File: src/main/java/org/rumbledb/compiler/StaticContextVisitor.java [hunk 3]
```diff
@@ -102,18 +112,31 @@ public StaticContext visitFunctionDeclaration(FunctionDeclaration declaration, S
...
+                mode = ExecutionMode.LOCAL;
+            }
+            functionDeclarationContext.addVariable(
+                name,
+                expression.getParams().get(name),
+                expression.getMetadata(),
+                mode
             );
+            ++i;
+        }
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/StaticContextVisitor.java:135]: Is it possible to extract this code block to a function?

File: src/main/java/org/rumbledb/compiler/VisitorConfig.java [hunk 4]
```diff
@@ -8,11 +8,14 @@
     private boolean suppressErrorsForCallingMissingFunctions;
     // flag to suppress errors when an unset execution mode value is fetched from a node
     private boolean suppressErrorsForAccessingUnsetExecutionModes;
+    // flag to suppress errors when an unset execution mode value is fetched from a node
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/VisitorConfig.java:11]: Copy pasted comment. Please remove or update

File: src/main/java/org/rumbledb/compiler/VisitorHelpers.java (hunk 1 above)
Comment (CanBerker) [reply src/main/java/org/rumbledb/compiler/VisitorHelpers.java:83]: Very nice catch!

File: src/main/java/org/rumbledb/compiler/VisitorConfig.java [hunk 5]
```diff
@@ -91,12 +108,14 @@ public boolean suppressErrorsForAccessingUnsetExecutionModes() {
         .withsuppressErrorsForFunctionSignatureCollision(true)
         .withSuppressErrorsForCallingMissingFunctions(false)
         .withSuppressErrorsForAccessingUnsetExecutionModes(false)
+        .withSetUnsetToLocal(true)
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/VisitorConfig.java:111]: Previously, we would set all unset function identifiers to local(without using a visitor) to solve deadlocks and then make a final visitor pass.

I wanted to clarify the current state after seeing this new flag `setUnsetToLocal`:
It seems that this flag is only used to set unset variable references to local. If this is the case I would recommend updating the name to indicate this. 

Furthermore, (I feel my brain is a bit bloated with information so I may be missing some important details while asking the following two questions):
- Do we still set the functions from Unset to Local "prior" to the final pass? Or do we do it during the final pass?
- In what scenerio do variables require a forced execution more from unset to local?

File: src/main/java/org/rumbledb/compiler/VisitorHelpers.java (hunk 1 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/VisitorHelpers.java:83]: Thanks!

File: src/main/java/org/rumbledb/compiler/VisitorHelpers.java (hunk 2 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/VisitorHelpers.java]: Good catch, done.

File: src/main/java/org/rumbledb/compiler/StaticContextVisitor.java (hunk 3 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/StaticContextVisitor.java:135]: Good idea, this is now separated.

File: src/main/java/org/rumbledb/compiler/VisitorConfig.java (hunk 4 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/VisitorConfig.java:11]: done

File: src/main/java/org/rumbledb/compiler/VisitorConfig.java (hunk 5 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/VisitorConfig.java]: > I wanted to clarify the current state after seeing this new flag `setUnsetToLocal`:
> It seems that this flag is only used to set unset variable references to local. If this is the case I would recommend updating the name to indicate this.

Yes, that is the case and this is renamed.

Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/VisitorConfig.java]: > * Do we still set the functions from Unset to Local "prior" to the final pass? Or do we do it during the final pass?

Yes, we still do. This code is still here.

**Comment by ghislainfourny** [comment]
Thanks a lot for looking, Can.

There are still many many things that I want to improve in here, so this is a small step that fits in the bigger picture.

**Comment by CanBerker** [comment]
The recent commits look great. I am approving the PR. When you make the tests succeed, feel free to merge.

**Pull request merged**